cleanlab_studio/internal/api/api.py

```diff
@@ -329,3 +333,64 @@ def poll_progress(
             res = request_function(progress_id)
         pbar.update(float(1) - pbar.n)
     return res
+
+
+def upload_predict_batch(api_key: str, model_id: str, batch: io.StringIO) -> str:
+    """Uploads prediction batch and returns query ID."""
+    url = f"{model_base_url}/{model_id}/upload"
+    res = requests.post(
+        url,
+        headers=_construct_headers(api_key),
+    )
+
+    handle_api_error(res)
+    presigned_url = res.json()["upload_url"]
+    query_id: str = res.json()["query_id"]
+    modality = res.json()["modality"]
+    header = res.json()["header"]
+    if modality == "text":
+        header_io = io.StringIO(header)
+        input_batch = io.StringIO("\n".join(chain(header_io, batch)))
+    else:
+        input_batch = batch
+
+    requests.post(presigned_url["url"], data=presigned_url["fields"], files={"file": input_batch})
+
+    return query_id
+
+
+def start_prediction(api_key: str, query_id: str) -> None:
+    """Starts prediction for query."""
+    res = requests.post(
+        f"{model_base_url}/predict/{query_id}",
+        headers=_construct_headers(api_key),
+    )
+
+    handle_api_error(res)
+
+
+def get_prediction_status(api_key: str, query_id: str) -> Dict[str, str]:
+    """Gets status of model prediction query."""
+    res = requests.get(
+        f"{model_base_url}/predict/{query_id}",
+        headers=_construct_headers(api_key),
+    )
+    handle_api_error(res)
+
+    prediction_results = res.json()
+    status = prediction_results["status"]
+    result_url = prediction_results["results"]
+    error_msg = prediction_results["error_msg"]
+
+    if status == "COMPLETE":
+        return {"status": "done", "result_url": result_url}
+    elif status == "FAILED":
+        return {"status": "error", "error_msg": error_msg}
+    else:
+        return {"status": "running"}
+
+
+def download_prediction_results(result_url: str) -> io.StringIO:
+    """Downloads prediction results from presigned URL."""
+    res = requests.get(result_url)
+    return io.StringIO(res.text)
```

Comment:
no need to do this buffering, just pass `res.raw`

Reply:
I think we do need this buffering since we want to return a numpy array for the results to the user.